clients/ruby/lib/samsara_sdk/ring_buffer.rb

```diff
@@ -49,36 +54,71 @@ def push(value)
     # @yieldparam [Array<Object>] data Buffer data.
     # @yieldreturn [Boolean] Result of data processing. True if success, false otherwise.
     #
-    # @return [Array<Object>] All buffer's elements.
+    # @return [Array<Object>] All buffer's elements in FIFO order.
     def flush
-      data, snapshot = take_snapshot
+      data, at_mark = take_snapshot
       success = block_given? ? yield(data) : TRUE
-      delete snapshot if success
+      delete at_mark if success
       data
     end
 
     private
 
-    # Get all actual elements from buffer at a given moment in FIFO order.
-    # Make a snapshot of the current buffer.
+    # Helper-method for calculating position in a circle.
+    #
+    # @return [Integer] position of a given pointer.
+    def calculate_position(pointer)
+      @size.nonzero? ? pointer.remainder(@size) : -1
+    end
+
+    # Get array-position of the element that was last added.
+    # Must be used within a synced mutex.
+    #
+    # @return [Integer] position.
+    def high_position
+      calculate_position @high
+    end
+
+    # Get array-position of the element that was last consumed.
+    # Must be used within a synced mutex.
+    #
+    # @return [Integer] position.
+    def low_position
+      calculate_position @low
+    end
+
+    # Make a snapshot of the current buffer at a given moment in FIFO order.
     #
     # @return [Array<Object>] actual elements in FIFO order,
-    # @return [Array<Object>] snapshot of current buffer.
+    # @return [Integer] High-water-mark that snapshot was taken at.
     def take_snapshot
       @mutex.synchronize do
-        i = @pointer == @size - 1 ? 0 : @pointer + 1
-        return (@buffer[i..-1] + @buffer[0...i]).compact, @buffer.dup
+        i = high_position
+        j = low_position + 1
+
+        data = if empty?
+                 []
+               elsif @low == -1
+                 @buffer.compact
+               elsif @high < @size # array is not filled completely
+                 @buffer[j..i]
+               else
+                 @buffer[j..-1] + @buffer[0..i]
+               end
+
+        return data, @high
```

Comment:
got it, now I understand why you did so.
I generally prefer clear code over performances unless performances are a clear project constraint which limit the outcome.
However in this case the performance "hack" is limited to a single function,
so if we document well this one explaining the motivation we can leave it as it is.

We can add a comment like:
```
take_snapshot returns a new array of all elements between @low and @high copying each element. Rater than copying the elements via Ruby's Array facility we use the lower and
more efficient C array copy.
etc etc  
```